src/vsr/replica.zig

```diff
@@ -731,7 +731,7 @@ pub fn ReplicaType(
                 assert(trailer_chunks.len == 1);
                 assert(trailer_size == ClientSessions.encode_size);
                 assert(trailer_size == trailer_chunks[0].len);
-
+                assert(self.client_replies.writing.count() == 0);
```

Comment:
@sentientwaffle how do `client_replies` interact with state sync?  In `read_reply_sync` we have

```zig
assert(write_latest.?.message.header.checksum == session.header.checksum);
```

That is, we assume that in-flight replies are always consistent with the session we have. So it feels like we need this assertion, to make sure we swap sessions when client_replies is quiescent.

_But_ I don't see any code which waits until `client_replies` completes all of its writes, so it feels that this assert can actually trigger?

Reply:
:thinking: I think you're right. `ClientReplies.checkpoint()` _does_ wait until all reply writes are flushed, but we are missing a `ClientReplies.cancel()` to do likewise before we state sync.

Reply:
>we are missing a ClientReplies.cancel() to do likewise

On the closer reading, it seems that's actually intentionally concurrent: 

```
            // Clear the fault *before* the write completes, not after.
            // Otherwise, a replica exiting state sync might mark a reply as faulty, then the
            // ClientReplies clears that bit due to an unrelated write that was already queued.
            client_replies.faulty.unset(slot.index);
```

which feels fine, and probably even more reliable than an explicit cancellation, but, yeah, that logic requires accounting for a case of a wrong reply in `read_reply_sync` as well.